<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en-us">
<head>
  <link href="http://gmpg.org/xfn/11" rel="profile">
  <meta http-equiv="content-type" content="text/html; charset=utf-8">

  
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1">

  <meta property="og:title" content="An introduction to javascript promises" />
<meta property="og:description" content="" />

<meta property="og:type" content="article" />

<meta property="og:locale" content="en_US" />
<meta property="og:url" content="https://vaneyckt.io/posts/an_introduction_to_javascript_promises/" />


  <title> An introduction to javascript promises  &middot; vaneyckt.io </title>

  

  <link rel="stylesheet" href="/css/monokai.css">
  <script src="/js/highlight.pack.js"></script>

  <script>hljs.initHighlightingOnLoad();</script>

  
  <link rel="stylesheet" href="https://vaneyckt.io/css/poole.css">
  <link rel="stylesheet" href="https://vaneyckt.io/css/syntax.css">
  <link rel="stylesheet" href="https://vaneyckt.io/css/hyde.css">
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=PT+Sans:400,400italic,700|Abril+Fatface">

  
  <link rel="apple-touch-icon-precomposed" sizes="144x144" href="/apple-touch-icon-144-precomposed.png">
  <link rel="shortcut icon" href="/favicon.ico">

  
  <link href="" rel="alternate" type="application/rss+xml" title="vaneyckt.io" />

  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.5.0/css/font-awesome.min.css">

  <link href='https://fonts.googleapis.com/css?family=Raleway:400,300' rel='stylesheet' type='text/css'>

  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-71853042-1', 'auto');
    ga('send', 'pageview');
  </script>
</head>

<body>

<div class="sidebar">
  <div class="container sidebar-sticky">
    <div class="sidebar-about">
      <h1 class="brand"><a style="text-decoration:none" href="https://vaneyckt.io">vaneyckt</a></h1>
      <p class="lead">
         notes to my future self 
      </p>
    </div>

    <ul class="sidebar-nav">
      <li><a href="https://vaneyckt.io">Home</a></li>
      <li><a href="https://vaneyckt.io/posts">Posts</a></li>
      <li><a href="https://vaneyckt.io/topics">Tags</a></li>
      
      <br/>
      
    </ul>
      
      
      
      <a href="https://github.com/vaneyckt"><i class="fa fa-github-square"></i></a>&nbsp;&nbsp;
      <a href="mailto:tomvaneyck@gmail.com"><i class="fa fa-envelope-square"></i></a>&nbsp;&nbsp;
      <a href="https://vaneyckt.io/index.xml"><i class="fa fa-rss-square"></i></a>&nbsp;&nbsp;
      

    <p class="footnote">powered by <a href="http://hugo.spf13.com">Hugo</a> <br/>
    &copy; 2016 Tom Van Eyck. All rights reserved.</p>
  </div>
</div>


  <div class="content container">
    <div class="post">
      <h1 class="post-title">An introduction to javascript promises</h1>
      <span class="post-date">Feb 7, 2015</span>
      

<p>I recently had to write some javascript code that required the sequential execution of half a dozen asynchronous requests. I figured this was the perfect time to learn a bit more about javascript promises. This post is a recap of what I read in these <a href="http://www.html5rocks.com/en/tutorials/es6/promises/">three</a> <a href="http://www.mullie.eu/how-javascript-promises-work/">amazing</a> <a href="http://www.sitepoint.com/overview-javascript-promises/">write-ups</a>.</p>

<h3 id="what-are-promises">What are promises?</h3>

<p>A Promise object represents a value that may not be available yet, but will be resolved at some point in future. This abstraction allows you to write asynchronous code in a more synchronous fashion. For example, you can use a Promise object to represent data that will eventually be returned by a call to a remote web service. The <code>then</code> and <code>catch</code> methods can be used to attach callbacks that will be triggered once the data arrives. We&rsquo;ll take a closer look at these two methods in the next sections. For now, let&rsquo;s write a simple AJAX request example that prints a random joke.</p>

<pre><code class="language-javascript">var promise = new Promise(function(resolve, reject) {
  $.ajax({
    url: &quot;http://api.icndb.com/jokes/random&quot;,
    success: function(result) {
      resolve(result[&quot;value&quot;][&quot;joke&quot;]);
    }
  });
});

promise.then(function(result) {
  console.log(result);
});
</code></pre>

<p>Note how the Promise object is just a wrapper around the AJAX request and how we&rsquo;ve instructed the <code>success</code> callback to trigger the <code>resolve</code> method. We&rsquo;ve also attached a callback to our Promise object with the <code>then</code> method. This callback gets triggered when the <code>resolve</code> method gets called. The <code>result</code> variable of this callback will contain the data that was passed to the <code>resolve</code> method.</p>

<p>Before we take a closer look at the <code>resolve</code> method, let&rsquo;s first investigate the Promise object a bit more. A Promise object can have one of three states:</p>

<ul>
<li><strong>fulfilled</strong> - the action relating to the Promise succeeded</li>
<li><strong>rejected</strong> - the action relating to the Promise failed</li>
<li><strong>pending</strong> - the Promise hasn&rsquo;t been fulfilled or rejected yet</li>
</ul>

<p>A pending Promise object can be fulfilled or rejected by calling <code>resolve</code> or <code>reject</code> on it. Once a Promise is fulfilled or rejected, this state gets permanently associated with it. The state of a fulfilled Promise also includes the data that was passed to <code>resolve</code>, just as the state of a rejected Promise also includes the data that was passed to <code>reject</code>. In summary, we can say that a Promise executes only once and stores the result of its execution.</p>

<pre><code class="language-javascript">var promise = new Promise(function(resolve, reject) {
  $.ajax({
    url: &quot;http://api.icndb.com/jokes/random&quot;,
    success: function(result) {
      resolve(result[&quot;value&quot;][&quot;joke&quot;]);
    }
  });
});

promise.then(function(result) {
  console.log(result);
});

promise.then(function(result) {
  console.log(result);
});
</code></pre>

<p>We can test whether a Promise only ever executes once by adding a second callback to the previous example. In this case, we see that only one AJAX request gets made and that the same joke gets printed to the console twice. This clearly shows that our Promise was only executed once.</p>

<h3 id="the-then-method-and-chaining">The <code>then</code> method and chaining</h3>

<p>The <code>then</code> method takes two arguments: a mandatory success callback and an optional failure callback. These callbacks are called when the Promise is settled (i.e. either fulfilled or rejected). If the Promise was fulfilled, the success callback will be fired with the data you passed to <code>resolve</code>. If the Promise was rejected, the failure callback will be called with the data you passed to <code>reject</code>. We&rsquo;ve already covered most of this in the previous section.</p>

<p>The real magic with the <code>then</code> method happens when you start chaining several of them together. This chaining allows you to express your logic in separate stages, each of which can be made responsible for transforming data passed on by the previous stage or for running additional asynchronous requests. The code below shows how data returned by the success callback of the first <code>then</code> method becomes available to the success callback of the second <code>then</code> method.</p>

<pre><code class="language-javascript">var promise = new Promise(function(resolve, reject) {
  $.ajax({
    url: &quot;http://api.icndb.com/jokes/random&quot;,
    success: function(result) {
      resolve(result[&quot;value&quot;][&quot;joke&quot;]);
    }
  });
});

promise.then(function(result) {
  return result;
}).then(function(result) {
  console.log(result);
});
</code></pre>

<p>This chaining is possible because the <code>then</code> method returns a new Promise object that will resolve to the return value of the callback. Or in other words, by calling <code>return result;</code> we cause the creation of an anonymous Promise object that looks something like shown below. Notice that this particular anonymous Promise object will resolve immediately, as it does not make any asynchronous requests.</p>

<pre><code class="language-javascript">new Promise(function(resolve, reject) {
  resolve(result);
});
</code></pre>

<p>Now that we understand that the <code>then</code> method always returns a Promise object, let&rsquo;s take a look at what happens when we tell the callback of a <code>then</code> method to explicitly return a Promise object.</p>

<pre><code class="language-javascript">function getJokePromise() {
  return new Promise(function(resolve, reject) {
    $.ajax({
      url: &quot;http://api.icndb.com/jokes/random&quot;,
      success: function(result) {
        resolve(result[&quot;value&quot;][&quot;joke&quot;]);
      }
    });
  });
}

getJokePromise().then(function(result) {
  console.log(result);
  return getJokePromise();
}).then(function(result) {
  console.log(result);
});
</code></pre>

<p>In this case, we end up sequentially executing two asynchronous requests. When the first Promise is resolved, the first joke is printed and a new Promise object is returned by the <code>then</code> method. This new Promise object then has <code>then</code> called on it. When the Promise succeeds, the <code>then</code> success callback is triggered and the second joke is printed.</p>

<p>The takeaway from all this is that calling <code>return</code> in a <code>then</code> callback will always result in returning a Promise object. It is this that allows for <code>then</code> chaining!</p>

<h3 id="error-handling">Error handling</h3>

<p>We mentioned in the previous section how the <code>then</code> method can take an optional failure callback that gets triggered when <code>reject</code> is called. It is customary to reject with an Error object as they capture a stack trace, thereby facilitating debugging.</p>

<pre><code class="language-javascript">var promise = new Promise(function(resolve, reject) {
  $.ajax({
    url: &quot;http://random.url.com&quot;,
    success: function(result) {
      resolve(result[&quot;value&quot;][&quot;joke&quot;]);
    },
    error: function(jqxhr, textStatus) {
      reject(Error(&quot;The AJAX request failed.&quot;));
    }
  });
});

promise.then(function(result) {
  console.log(result);
}, function(error) {
  console.log(error);
  console.log(error.stack);
});
</code></pre>

<p>Personally, I find this a bit hard to read. Luckily we can use the <code>catch</code> method to make this look a bit nicer. There&rsquo;s nothing special about the <code>catch</code> method. In fact, it&rsquo;s just sugar for <code>then(undefined, func)</code>, but it definitely makes code easier to read.</p>

<pre><code class="language-javascript">var promise = new Promise(function(resolve, reject) {
  $.ajax({
    url: &quot;http://random.url.com&quot;,
    success: function(result) {
      resolve(result[&quot;value&quot;][&quot;joke&quot;]);
    },
    error: function(jqxhr, textStatus) {
      reject(Error(&quot;The AJAX request failed.&quot;));
    }
  });
});

promise.then(function(result) {
  console.log(result);
}).then(function(result) {
  console.log(&quot;foo&quot;); // gets skipped
}).then(function(result) {
  console.log(&quot;bar&quot;); // gets skipped
}).catch(function(error) {
  console.log(error);
  console.log(error.stack);
});
</code></pre>

<p>Aside from illustrating improved readability, the above code showcases another aspect of the <code>reject</code> method in that Promise rejections will cause your code to skip forward to the next <code>then</code> method that has a rejection callback (or the next <code>catch</code> method, since this is equivalent). It is this fallthrough behavior that causes this code to not print &ldquo;foo&rdquo; or &ldquo;bar&rdquo;!</p>

<p>As a final point, it is useful to know that a Promise is implicitly rejected if an error is thrown in its constructor callback. This means it&rsquo;s useful to do all your Promise related work inside the Promise constructor callback, so errors automatically become rejections.</p>

<pre><code class="language-javascript">var promise = new Promise(function(resolve, reject) {
  // JSON.parse throws an error if you feed it some
  // invalid JSON, so this implicitly rejects
  JSON.parse(&quot;This ain't JSON&quot;);
});

promise.then(function(result) {
  console.log(result);
}).catch(function(error) {
  console.log(error);
});
</code></pre>

<p>The above code will cause the Promise to be rejected and an error to be printed because it will fail to parse the invalid JSON string.</p>

    </div>
  </div>

</body>
</html>
